clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -944,22 +951,259 @@ static llvm::Value *getArrayIndexingBound(CodeGenFunction &CGF,
+      }
+    }
+
+    if (const auto *Record = dyn_cast<RecordDecl>(D)) {
+      ++FieldNo;
+      if (getGEPIndicesToField(CGF, Record, FD, Indices)) {
+        Indices.emplace_back(std::make_pair(RD, CGF.Builder.getInt32(FieldNo)));
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+llvm::Value *CodeGenFunction::EmitCountedByFieldExpr(
+    const Expr *Base, const FieldDecl *FAMDecl, const FieldDecl *CountDecl) {
+  // This method is typically called in contexts where we can't generate
+  // side-effects, like in __builtin_dynamic_object_size. When finding
+  // expressions, only choose those that have either already been emitted or
+  // can be loaded without side-effects.
+  const RecordDecl *RD = CountDecl->getParent()->getOuterLexicalRecordContext();
+
+  // Find the base struct expr (i.e. p in p->a.b.c.d).
+  const Expr *StructBase = StructAccessBase(RD).Visit(Base);
+  if (!StructBase)
+    return nullptr;
+
+  llvm::Value *Res = nullptr;
+  if (const auto *DRE = dyn_cast<DeclRefExpr>(StructBase)) {
+    Res = EmitDeclRefLValue(DRE).getPointer(*this);
+
+    const Type *Ty = DRE->getType().getTypePtr();
+    Res = Builder.CreateAlignedLoad(ConvertType(QualType(Ty, 0)), Res,
+                                    getPointerAlign(), "dre.load");
+  } else {
+    Address Addr = Address::invalid();
+    auto I = LocalDeclMap.find(FAMDecl);
+    if (I != LocalDeclMap.end()) {
+      Res = I->second.getPointer();
+      Addr = Address(Res, ConvertType(FAMDecl->getType()), getPointerAlign());
+    }
+
+    bool NeedLoad = true;
+    if (!Res && !StructBase->HasSideEffects(getContext())) {
+      LValueBaseInfo EltBaseInfo;
+      TBAAAccessInfo EltTBAAInfo;
+      Addr = EmitPointerWithAlignment(StructBase, &EltBaseInfo, &EltTBAAInfo);
+      NeedLoad = false;
+    }
+
+    if (!Addr.isValid())
+      return nullptr;
+
+    if (NeedLoad)
+      Res = Builder.CreateLoad(Addr, /*isVolatile=*/false, "struct.load");
+
+    Res = Addr.getPointer();
+  }
+
+  if (!Res)
+    return nullptr;
+
+  llvm::Value *Zero = Builder.getInt32(0);
+  RecIndicesTy Indices;
+
+  getGEPIndicesToField(*this, RD, CountDecl, Indices);
+
+  for (auto I = Indices.rbegin(), E = Indices.rend(); I != E; ++I)
+    Res = Builder.CreateInBoundsGEP(
+        ConvertType(QualType(I->first->getTypeForDecl(), 0)), Res,
+        {Zero, I->second}, "..counted_by.gep");
+
+  return Builder.CreateAlignedLoad(ConvertType(CountDecl->getType()), Res,
+                                   getIntAlign(), "..counted_by.load");
+}
+
+const FieldDecl *CodeGenFunction::FindCountedByField(const FieldDecl *FD) {
+  if (!FD || !FD->hasAttr<CountedByAttr>())
+    return nullptr;
+
+  const auto *CBA = FD->getAttr<CountedByAttr>();
+  if (!CBA)
+    return nullptr;
+
+  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
+    while (RD && !RD->getDeclName())
+      if (const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext()))
+        RD = R;
+      else
+        break;
```

Comment:
I changed this to:

```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD && RD->isAnonymousStructOrUnion()) {
      const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext());
      if (!R)
        break;
      RD = R;
    }
    return RD;
  };
```

I must not have pushed the change up. This is another instance of your above question about whether `R` could ever be `nullptr`. If it can't be `nullptr`, I can change this into:

```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD->isAnonymousStructOrUnion())
      RD = cast<RecordDecl>(RD->getDeclContext());
    return RD;
  };
```

Is that an okay assumption to make here?

Reply:
It looks like I can get away with a modified form of the previous version. Done.

Reply:
But in the modified form, wouldn't `RD = cast<RecordDecl>(RD->getDeclContext());` crash if it can be a non record decl as you explained with the example?

I wonder if it should just be something like:
```
  auto GetNonAnonStructOrUnion = [](const RecordDecl *RD) {
    while (RD && RD->isAnonymousStructOrUnion()) {
      const auto *R = dyn_cast<RecordDecl>(RD->getDeclContext());
      if (!R)
        return nullptr; // <- THIS
      RD = R;
    }
    return RD;
  };

Reply:
Ah I see your new commit does this already.